Sources/MapboxDirections/RouteResponse.swift

```diff
@@ -91,37 +91,63 @@ extension RouteResponse: Codable {
-            optionsWaypoints = matchOpts.waypoints
+            optionsWaypoints = matchOpts.waypoints.map {
+                Route.Waypoint(coordinate: $0.coordinate,
+                               correction: 0,
+                               name: $0.name)
+            }
         case let .route(options: routeOpts):
-            optionsWaypoints = routeOpts.waypoints
+            optionsWaypoints = routeOpts.waypoints.map {
+                Route.Waypoint(coordinate: $0.coordinate,
+                               correction: 0,
+                               name: $0.name)
+            }
         }
                 
         if let decodedWaypoints = decodedWaypoints {
             // The response lists the same number of tracepoints as the waypoints in the request, whether or not a given waypoint is leg-separating.
-            waypoints = zip(decodedWaypoints, optionsWaypoints).map { (pair) -> Waypoint in
-                let (decodedWaypoint, waypointInOptions) = pair
-                let waypoint = Waypoint(coordinate: decodedWaypoint.coordinate, coordinateAccuracy: waypointInOptions.coordinateAccuracy, name: waypointInOptions.name?.nonEmptyString ?? decodedWaypoint.name)
-                waypoint.separatesLegs = waypointInOptions.separatesLegs
-                return waypoint
+            waypoints = zip(decodedWaypoints, optionsWaypoints).map { (pair) -> Route.Waypoint in
+                var (decodedWaypoint, waypointInOptions) = pair
+                if let name = waypointInOptions.name?.nonEmptyString {
+                    decodedWaypoint.name = name
+                }
+                return decodedWaypoint
             }
-            waypoints?.first?.separatesLegs = true
-            waypoints?.last?.separatesLegs = true
         } else {
             waypoints = decodedWaypoints
         }
         
         if let routes = try container.decodeIfPresent([Route].self, forKey: .routes) {
             // Postprocess each route.
+            var legSeparators = waypoints?.compactMap { $0 }
+            if let options = decoder.userInfo[.options] as? DirectionsOptions, let waypoints = waypoints {
+                // Select first, last and all other waypoints which separate legs.
+                var optionsLegSeparators = options.legSeparators
+                legSeparators = zip(waypoints, options.waypoints).compactMap { (pair) -> Route.Waypoint? in
+                    let (decodedWaypoint, waypointInOptions) = pair
+                    if decodedWaypoint == waypoints.last! ||
+                        decodedWaypoint == waypoints.first! {
+                        return decodedWaypoint
+                    }
+                    
+                    guard let index = optionsLegSeparators.firstIndex(of: waypointInOptions) else {
+                        return nil
+                    }
+                    optionsLegSeparators.remove(at: index)
+                    return decodedWaypoint
```

Comment:
I hope I understood correctly how filtering should be applied. I decided to stick to `options.legSeparators` and filter out decoded waypoints that correspond it.